File: internal/domain/service/payment_service.go
package service

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"time"

	"github.com/google/uuid"

	"foto-app/internal/domain/model"
	"foto-app/internal/domain/port"
)

// ---------- Request/Response types ----------

// InitiatePaymentResponse carries the result of initiating a payment.
type InitiatePaymentResponse struct {
	Payment   *model.Payment `json:"payment"`
	PaymentURL *string       `json:"payment_url,omitempty"` // MP init_point
}

// MPWebhookPayload represents the incoming Mercado Pago webhook body.
type MPWebhookPayload struct {
	Action   string          `json:"action"`
	Type     string          `json:"type"`
	Data     *MPWebhookData  `json:"data"`
	ID       json.Number     `json:"id"` // notification ID
}

// MPWebhookData holds the data.id field from the webhook.
type MPWebhookData struct {
	ID string `json:"id"`
}

// ---------- Payment Service ----------

// PaymentService implements business logic for payment processing.
// It coordinates between PaymentRepository, OrderRepository, and PaymentGateway
// to handle MP Checkout Pro payments and cash-on-delivery payments.
type PaymentService struct {
	paymentRepo port.PaymentRepository
	orderRepo   port.OrderRepository
	gateway     port.PaymentGateway
}

// NewPaymentService creates a PaymentService with its dependencies.
func NewPaymentService(
	paymentRepo port.PaymentRepository,
	orderRepo port.OrderRepository,
	gateway port.PaymentGateway,
) *PaymentService {
	return &PaymentService{
		paymentRepo: paymentRepo,
		orderRepo:   orderRepo,
		gateway:     gateway,
	}
}

// InitiatePayment starts the payment flow for an order.
//
// For "mercadopago": creates an MP Checkout Pro preference, stores the
// preference ID on the order, and returns the init_point URL.
//
// For "cash": creates a pending payment record. An admin must later
// confirm the payment via ConfirmCashPayment.
//
// Returns ErrOrderAlreadyPaid if the order has already been paid.
// Returns ErrInvalidPayment if the method does not match the order's method.
// Returns ErrNotFound if the order does not exist or doesn't belong to the student.
func (s *PaymentService) InitiatePayment(
	ctx context.Context,
	studentID uuid.UUID,
	orderID uuid.UUID,
	method string,
) (*InitiatePaymentResponse, error) {
	// 1. Find order
	order, err := s.orderRepo.FindByID(ctx, orderID)
	if err != nil {
		return nil, err
	}
	if order.StudentID != studentID {
		return nil, model.ErrNotFound
	}

	// 2. Validate payment method
	if order.PaymentMethod != method {
		return nil, model.ErrInvalidPayment
	}

	// 3. Check not already paid
	if order.PaymentStatus == model.PaymentStatusPaid {
		return nil, model.ErrOrderAlreadyPaid
	}

	// 4. Check existing payment record
	existingPayment, err := s.paymentRepo.FindByOrderID(ctx, orderID)
	if err != nil && !errors.Is(err, model.ErrNotFound) {
		return nil, err
	}
	if existingPayment != nil && existingPayment.Status == model.PaymentStatusApproved {
		return nil, model.ErrOrderAlreadyPaid
	}
	if existingPayment != nil && existingPayment.Status != model.PaymentStatusPending {
		// Payment exists in a terminal state — reject
		return nil, model.ErrOrderAlreadyPaid
	}

	switch method {
	case model.PaymentMethodMercadoPago:
		return s.initiateMP(ctx, order)
	case model.PaymentMethodCash:
		return s.initiateCash(ctx, order)
	default:
		return nil, model.ErrInvalidPayment
	}
}

func (s *PaymentService) initiateMP(ctx context.Context, order *model.Order) (*InitiatePaymentResponse, error) {
	// Get order items for MP preference
	items, err := s.orderRepo.FindItemsByOrderID(ctx, order.ID)
	if err != nil {
		return nil, fmt.Errorf("get order items: %w", err)
	}

	// Build MP preference items
	mpItems := make([]port.MPPreferenceItem, len(items))
	for i, it := range items {
		mpItems[i] = port.MPPreferenceItem{
			Title:      it.Title,
			Quantity:   it.Quantity,
			UnitPrice:  it.UnitPrice,
			CurrencyID: "ARS",
		}
	}

	// Create MP Checkout Pro preference
	backURLs := &port.MPBackURLs{
		Success: "",
		Failure: "",
		Pending: "",
	}
	pref, err := s.gateway.CreatePreference(ctx, order.ID.String(), mpItems, backURLs)
	if err != nil {
		return nil, errors.Join(model.ErrPaymentFailed, err)
	}

	// Persist payment record
	payment := &model.Payment{
		ID:                uuid.New(),
		OrderID:           order.ID,
		Method:            model.PaymentMethodMercadoPago,
		Status:            model.PaymentStatusPending,
		Amount:            order.Total,
		ExternalReference: strPtr(order.ID.String()),
		MPPaymentID:       &pref.ID,
	}
	if err := s.paymentRepo.Create(ctx, payment); err != nil {
		return nil, fmt.Errorf("create payment: %w", err)
	}

	// Store mp_preference_id on the order
	if err := s.orderRepo.SetMPPreferenceID(ctx, order.ID, pref.ID); err != nil {
		return nil, fmt.Errorf("set mp_preference_id: %w", err)
	}

	return &InitiatePaymentResponse{
		Payment:    payment,
		PaymentURL: &pref.InitPoint,
	}, nil
}

func (s *PaymentService) initiateCash(ctx context.Context, order *model.Order) (*InitiatePaymentResponse, error) {
	payment := &model.Payment{
		ID:                uuid.New(),
		OrderID:           order.ID,
		Method:            model.PaymentMethodCash,
		Status:            model.PaymentStatusPending,
		Amount:            order.Total,
		ExternalReference: strPtr(order.ID.String()),
	}

	if err := s.paymentRepo.Create(ctx, payment); err != nil {
		return nil, fmt.Errorf("create cash payment: %w", err)
	}

	return &InitiatePaymentResponse{
		Payment: payment,
	}, nil
}

// HandleMPWebhook processes an incoming Mercado Pago webhook notification.
//
// It is idempotent: if the same notification ID has already been processed,
// this is a no-op. The upsert uses the payment_events.event_id (MP notification ID)
// as the unique constraint.
//
// When processing a new "payment" event, it:
//  1. Queries MP API for the payment's current status
//  2. Finds the corresponding payment record by external_reference (order_id)
//  3. Updates the payment and order statuses accordingly
//
// Approved → order status: confirmed, payment_status: paid
// Rejected → payment_status: failed
func (s *PaymentService) HandleMPWebhook(ctx context.Context, rawBody json.RawMessage) error {
	var payload MPWebhookPayload
	if err := json.Unmarshal(rawBody, &payload); err != nil {
		return fmt.Errorf("parse webhook payload: %w", err)
	}

	if payload.Data == nil || payload.Data.ID == "" {
		return fmt.Errorf("missing data.id in webhook payload")
	}

	notificationID := payload.ID.String()
	if notificationID == "" || notificationID == "0" {
		// Generate a fallback ID if MP didn't provide one
		notificationID = payload.Data.ID + "-" + payload.Action
	}

	// Create event for idempotency check
	mpPaymentID := payload.Data.ID
	event := &model.PaymentEvent{
		ID:          uuid.New(),
		EventID:     notificationID,
		Topic:       payload.Type,
		Action:      payload.Action,
		MPPaymentID: &mpPaymentID,
		RawBody:     rawBody,
	}

	isNew, err := s.paymentRepo.UpsertEvent(ctx, event)
	if err != nil {
		return fmt.Errorf("upsert payment event: %w", err)
	}
	if !isNew {
		// Already processed — idempotent
		return nil
	}

	// Only process "payment" topic events
	if payload.Type != "payment" {
		return nil
	}

	// Query MP for payment status
	// The data.id might be a string that represents an integer
	var paymentID int
	if _, err := fmt.Sscanf(mpPaymentID, "%d", &paymentID); err != nil {
		return fmt.Errorf("parse mp payment id %q: %w", mpPaymentID, err)
	}

	mpInfo, err := s.gateway.GetPaymentInfo(ctx, paymentID)
	if err != nil {
		return fmt.Errorf("get mp payment info: %w", err)
	}

	// Find payment record by external_reference (order_id)
	payments, err := s.findPaymentByExternalRef(ctx, mpInfo.ExternalReference)
	if err != nil {
		return fmt.Errorf("find payment by external ref: %w", err)
	}
	if payments == nil {
		// Payment not found — maybe initiated outside our system
		return nil
	}

	// Map MP status to our statuses
	var paymentStatus string
	var orderStatus string
	var paidAt *time.Time

	switch mpInfo.Status {
	case "approved":
		paymentStatus = model.PaymentStatusApproved
		orderStatus = model.OrderStatusConfirmed
		now := time.Now()
		paidAt = &now
	case "rejected", "cancelled":
		paymentStatus = model.PaymentStatusRejected
		orderStatus = model.OrderStatusPending
		paidAt = nil
	case "refunded":
		paymentStatus = model.PaymentStatusRefunded
		orderStatus = model.OrderStatusConfirmed // order stays confirmed, just payment refunded
		paidAt = nil
	default:
		// "pending" or "in_process" — leave as-is
		return nil
	}

	// Update payment record
	if err := s.paymentRepo.UpdateStatus(ctx, payments.ID, paymentStatus, paidAt); err != nil {
		return fmt.Errorf("update payment status: %w", err)
	}

	if payments.OrderID != uuid.Nil {
		// Update order statuses
		if err := s.orderRepo.UpdatePaymentInfo(ctx, payments.OrderID, mapMpPaymentStatus(paymentStatus), orderStatus); err != nil {
			return fmt.Errorf("update order payment info: %w", err)
		}
	}

	return nil
}

// mapMpPaymentStatus converts our payment status to the order's payment_status enum.
// Both use the same values ("paid", "failed", "refunded"), just with different naming in the enum.
// The order's payment_status enum values are: 'pending', 'paid', 'failed', 'refunded'.
func mapMpPaymentStatus(status string) string {
	switch status {
	case model.PaymentStatusApproved:
		return model.PaymentStatusPaid
	case model.PaymentStatusRejected:
		return model.PaymentStatusFailed
	case model.PaymentStatusRefunded:
		return model.PaymentStatusRefunded
	default:
		return model.PaymentStatusPending
	}
}

// findPaymentByExternalRef finds a payment record by external_reference (order ID).
func (s *PaymentService) findPaymentByExternalRef(ctx context.Context, extRef string) (*model.Payment, error) {
	if extRef == "" {
		return nil, nil
	}
	orderID, err := uuid.Parse(extRef)
	if err != nil {
		return nil, nil
	}
	payment, err := s.paymentRepo.FindByOrderID(ctx, orderID)
	if err != nil {
		if errors.Is(err, model.ErrNotFound) {
			return nil, nil
		}
		return nil, err
	}
	return payment, nil
}

// ConfirmCashPayment marks a cash payment as paid by an admin.
//
// Validates that:
//   - The order exists
//   - The payment method is "cash"
//   - The payment is still pending
//
// On success, updates both the payment record and the order statuses.
func (s *PaymentService) ConfirmCashPayment(ctx context.Context, orderID uuid.UUID) error {
	// Find payment record
	payment, err := s.paymentRepo.FindByOrderID(ctx, orderID)
	if err != nil {
		return err
	}

	if payment.Method != model.PaymentMethodCash {
		return model.ErrNotCashPayment
	}

	if payment.Status != model.PaymentStatusPending {
		return model.ErrOrderAlreadyPaid
	}

	// Update payment
	now := time.Now()
	if err := s.paymentRepo.UpdateStatus(ctx, payment.ID, model.PaymentStatusApproved, &now); err != nil {
		return fmt.Errorf("update payment status: %w", err)
	}

	// Update order
	if err := s.orderRepo.UpdatePaymentInfo(ctx, orderID, model.PaymentStatusPaid, model.OrderStatusConfirmed); err != nil {
		return fmt.Errorf("update order payment info: %w", err)
	}

	return nil
}

// strPtr returns a pointer to the given string.
func strPtr(s string) *string {
	return &s
}

File: backend/src/services/payment.service.js
import { v4 as uuidv4 } from 'uuid';
import { prisma } from '../lib/prisma.js';
import { config } from '../config.js';

export class PaymentService {
  constructor(gateway) {
    this.gateway = gateway;
  }

  async initiatePayment(studentId, orderId, method) {
    const order = await prisma.order.findUnique({ where: { id: orderId } });
    if (!order) {
      const err = new Error('order not found');
      err.code = 'INF_001';
      err.status = 404;
      throw err;
    }
    if (order.studentId !== studentId) {
      const err = new Error('order not found');
      err.code = 'INF_001';
      err.status = 404;
      throw err;
    }
    if (order.paymentMethod !== method) {
      const err = new Error('invalid payment method for this order');
      err.code = 'PAY_002';
      err.status = 400;
      throw err;
    }
    if (order.paymentStatus === 'paid') {
      const err = new Error('order already paid');
      err.code = 'PAY_003';
      err.status = 400;
      throw err;
    }

    // Check existing payment record
    const existingPayment = await prisma.payment.findUnique({ where: { orderId } });
    if (existingPayment && existingPayment.status === 'approved') {
      const err = new Error('order already paid');
      err.code = 'PAY_003';
      err.status = 400;
      throw err;
    }
    if (existingPayment && existingPayment.status !== 'pending') {
      const err = new Error('order already paid');
      err.code = 'PAY_003';
      err.status = 400;
      throw err;
    }

    if (method === 'mercadopago') {
      return this.initiateMP(order);
    }
    if (method === 'cash') {
      return this.initiateCash(order);
    }
    
    const err = new Error('invalid payment method');
    err.code = 'PAY_002';
    err.status = 400;
    throw err;
  }

  async initiateMP(order) {
    const items = await prisma.orderItem.findMany({ where: { orderId: order.id } });

    const mpItems = items.map((item) => ({
      title: item.title,
      quantity: item.quantity,
      unitPrice: Number(item.unitPrice) / 100,
      currencyId: 'ARS',
    }));

    const pref = await this.gateway.createPreference(order.id, mpItems, {
      success: `${config.frontendUrl}/orders`,
      failure: `${config.frontendUrl}/orders`,
      pending: `${config.frontendUrl}/orders`,
    });

    const payment = await prisma.payment.create({
      data: {
        id: uuidv4(),
        orderId: order.id,
        method: 'mercadopago',
        status: 'pending',
        amount: order.total,
        externalReference: order.id,
        mpPaymentId: pref.id,
      },
    });

    await prisma.order.update({
      where: { id: order.id },
      data: { mpPreferenceId: pref.id },
    });

    return { payment, paymentUrl: pref.initPoint };
  }

  async initiateCash(order) {
    const payment = await prisma.payment.create({
      data: {
        id: uuidv4(),
        orderId: order.id,
        method: 'cash',
        status: 'pending',
        amount: order.total,
        externalReference: order.id,
      },
    });

    return { payment };
  }

  async handleMPWebhook(rawBody) {
    let payload;
    try {
      payload = JSON.parse(rawBody);
    } catch {
      throw new Error('parse webhook payload failed');
    }

    if (!payload.data || !payload.data.id) {
      throw new Error('missing data.id in webhook payload');
    }

    const notificationId = payload.id ? String(payload.id) : `${payload.data.id}-${payload.action}`;
    if (!notificationId || notificationId === '0') {
      throw new Error('invalid notification ID');
    }

    // Upsert event for idempotency
    const existing = await prisma.paymentEvent.findUnique({
      where: { eventId: notificationId },
    });

    if (!existing) {
      await prisma.paymentEvent.create({
        data: {
          id: uuidv4(),
          eventId: notificationId,
          topic: payload.type,
          action: payload.action,
          mpPaymentId: payload.data.id,
          rawBody: rawBody,
          processed: false,
        },
      });
    } else {
      // Already processed — idempotent
      return;
    }

    // Only process "payment" topic events
    if (payload.type !== 'payment') return;

    // Query MP for payment status
    const mpPaymentId = parseInt(payload.data.id, 10);
    if (isNaN(mpPaymentId)) {
      throw new Error(`parse mp payment id "${payload.data.id}" failed`);
    }

    const mpInfo = await this.gateway.getPaymentInfo(mpPaymentId);
    if (!mpInfo.externalReference) return;

    const extRef = mpInfo.externalReference;

    // ── NEW FLOW: PendingCheckout (no order created yet) ────────────
    const pendingCheckout = await prisma.pendingCheckout.findUnique({
      where: { id: extRef },
    });

    if (pendingCheckout) {
      return this._handlePendingCheckoutWebhook(pendingCheckout, mpInfo);
    }

    // ── OLD FLOW: Existing order with a Payment record ──────────────
    const payment = await prisma.payment.findFirst({
      where: { externalReference: extRef },
    });
    if (!payment) return;

    // Map MP status
    let paymentStatus;
    let orderStatus;
    let paidAt = null;

    switch (mpInfo.status) {
      case 'approved':
        paymentStatus = 'approved';
        orderStatus = 'pending';
        paidAt = new Date();
        break;
      case 'rejected':
      case 'cancelled':
        paymentStatus = 'rejected';
        orderStatus = 'cancelled';
        // Restore stock when payment fails
        const failedItems = await prisma.orderItem.findMany({
          where: { orderId: payment.orderId },
        });
        for (const item of failedItems) {
          await prisma.booklet.update({
            where: { id: item.bookletId },
            data: { stock: { increment: item.quantity } },
          });
        }
        break;
      case 'refunded':
        paymentStatus = 'refunded';
        orderStatus = 'pending';
        break;
      default:
        return; // "pending" or "in_process" — leave as-is
    }

    // Map to order payment_status enum
    const orderPaymentStatus = this.mapMpPaymentStatus(paymentStatus);

    // Update payment
    await prisma.payment.update({
      where: { id: payment.id },
      data: { status: paymentStatus, paidAt },
    });

    // Update order
    if (payment.orderId) {
      await prisma.order.update({
        where: { id: payment.orderId },
        data: { paymentStatus: orderPaymentStatus, status: orderStatus },
      });
    }
  }

  // ── PendingCheckout webhook handling ─────────────────────────────────

  async _handlePendingCheckoutWebhook(pendingCheckout, mpInfo) {
    switch (mpInfo.status) {
      case 'approved':
        return this._completeCheckout(pendingCheckout);
      case 'rejected':
      case 'cancelled':
      case 'refunded':
        await prisma.pendingCheckout.update({
          where: { id: pendingCheckout.id },
          data: { status: 'failed' },
        });
        return;
      default:
        return; // "pending" or "in_process" — leave as-is
    }
  }

  async _completeCheckout(pendingCheckout) {
    const items = pendingCheckout.items;
    const studentId = pendingCheckout.studentId;
    const orderId = uuidv4();

    // 1. Re-validate stock BEFORE creating anything (outside transaction)
    for (const item of items) {
      const booklet = await prisma.booklet.findFirst({
        where: { id: item.bookletId },
        select: { stock: true, isActive: true },
      });
      if (!booklet || !booklet.isActive || booklet.stock < item.quantity) {
        await prisma.pendingCheckout.update({
          where: { id: pendingCheckout.id },
          data: { status: 'failed' },
        });
        console.error(
          `[MP Checkout] Stock shortage for booklet ${item.bookletId} ` +
          `(need ${item.quantity}, have ${booklet?.stock ?? 0}). ` +
          `PendingCheckout ${pendingCheckout.id} marked as failed. ` +
          `User was charged but no order created — needs admin refund.`
        );
        return;
      }
    }

    await prisma.$transaction(async (tx) => {
      // 2. Create order
      await tx.order.create({
        data: {
          id: orderId,
          studentId,
          total: pendingCheckout.total,
          status: 'pending',
          paymentMethod: 'mercadopago',
          paymentStatus: 'paid',
        },
      });

      // 3. Create order items + decrement stock + compute delivery
      let maxDeliveryDays = 0;

      for (const item of items) {
        await tx.orderItem.create({
          data: {
            id: uuidv4(),
            orderId,
            bookletId: item.bookletId,
            title: item.title,
            quantity: item.quantity,
            unitPrice: item.unitPrice,
            deliveryDays: item.deliveryDays,
          },
        });

        await tx.booklet.update({
          where: { id: item.bookletId },
          data: { stock: { decrement: item.quantity } },
        });

        if (item.deliveryDays && item.deliveryDays > maxDeliveryDays) {
          maxDeliveryDays = item.deliveryDays;
        }
      }

      // 4. Set delivery date
      if (maxDeliveryDays > 0) {
        await tx.order.update({
          where: { id: orderId },
          data: {
            deliveryDate: new Date(Date.now() + maxDeliveryDays * 24 * 60 * 60 * 1000),
          },
        });
      }

      // 5. Create payment record
      await tx.payment.create({
        data: {
          id: uuidv4(),
          orderId,
          method: 'mercadopago',
          status: 'approved',
          amount: pendingCheckout.total,
          externalReference: pendingCheckout.id,
          mpPaymentId: pendingCheckout.mpPreferenceId,
          paidAt: new Date(),
        },
      });

      // 6. Clear the student's cart
      const cart = await tx.cart.findFirst({ where: { studentId } });
      if (cart) {
        await tx.cartItem.deleteMany({ where: { cartId: cart.id } });
      }

      // 7. Mark pending checkout as completed
      await tx.pendingCheckout.update({
        where: { id: pendingCheckout.id },
        data: { status: 'completed' },
      });
    });
  }

  mapMpPaymentStatus(status) {
    switch (status) {
      case 'approved': return 'paid';
      case 'rejected': return 'failed';
      case 'refunded': return 'refunded';
      default: return 'pending';
    }
  }

  async confirmCashPayment(orderId) {
    const payment = await prisma.payment.findUnique({ where: { orderId } });
    if (!payment) {
      const err = new Error('order or payment not found');
      err.code = 'INF_001';
      err.status = 404;
      throw err;
    }
    if (payment.method !== 'cash') {
      const err = new Error('payment method is not cash');
      err.code = 'PAY_006';
      err.status = 400;
      throw err;
    }
    if (payment.status !== 'pending') {
      const err = new Error('payment already processed');
      err.code = 'PAY_003';
      err.status = 400;
      throw err;
    }

    const now = new Date();
    await prisma.payment.update({
      where: { id: payment.id },
      data: { status: 'approved', paidAt: now },
    });

    await prisma.order.update({
      where: { id: orderId },
      data: { paymentStatus: 'paid', status: 'pending' },
    });
  }
}
